src/Bicep.Core/Emit/ResourceDependencyVisitor.cs

```diff
@@ -9,37 +9,53 @@
-        private IDictionary<DeclaredSymbol, HashSet<DeclaredSymbol>> resourceDependencies;
-        private DeclaredSymbol? currentDeclaration;
+        // resource dependencies specific to each ResourceDependencyVisitor
+        private HashSet<DeclaredSymbol> resourceDependencies;
 
-        public static ImmutableDictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>> GetResourceDependencies(SemanticModel model)
+        public static ImmutableDictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>> GetAllResourceDependencies(SemanticModel model) 
         {
-            var visitor = new ResourceDependencyVisitor(model);
-            visitor.Visit(model.Root.Syntax);
+            var allResourceDependencies = new Dictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>>();
+            var declaredSymbols = model.Root.AllDeclarations.ToList(); // gets all declaredSymbols in the document (entry point)
+
+            // we need to precompute variable dependencies to reuse them when traversing resources or modules
+            foreach(var variableSymbol in declaredSymbols.OfType<VariableSymbol>()) 
+            {
+                variableDependencies[variableSymbol] = GetResourceDependencies(model, variableSymbol.DeclaringSyntax);
+            }
 
-            var output = new Dictionary<DeclaredSymbol, ImmutableHashSet<DeclaredSymbol>>();
-            foreach (var kvp in visitor.resourceDependencies)
+            foreach(var declaredSymbol in declaredSymbols.Where(symbol => symbol is ModuleSymbol || symbol is ResourceSymbol))
             {
-                if (kvp.Key is ResourceSymbol resourceSymbol)
-                {
-                    output[resourceSymbol] = kvp.Value.ToImmutableHashSet();
-                }
-                if (kvp.Key is ModuleSymbol moduleSymbol)
-                {
-                    output[moduleSymbol] = kvp.Value.ToImmutableHashSet();
-                }
+                allResourceDependencies[declaredSymbol] = GetResourceDependencies(model, declaredSymbol.DeclaringSyntax);
             }
-            return output.ToImmutableDictionary();
+            return allResourceDependencies.ToImmutableDictionary();
+        }
+
+        public static ImmutableHashSet<DeclaredSymbol> GetResourceDependencies(SemanticModel model, SyntaxBase syntax)
+        {
+            var visitor = new ResourceDependencyVisitor(model);
+            visitor.Visit(syntax);
+            return visitor.resourceDependencies.ToImmutableHashSet();
         }
 
         private ResourceDependencyVisitor(SemanticModel model)
         {
             this.model = model;
-            this.resourceDependencies = new Dictionary<DeclaredSymbol, HashSet<DeclaredSymbol>>();
-            this.currentDeclaration = null;
+            this.resourceDependencies = new HashSet<DeclaredSymbol>();
         }
 
         public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
```

Comment:
I think you should be able to get rid of the Visit*DeclarationSyntax methods now as they aren't doing anything.